### PR TITLE
2176 introduce attributes command

### PR DESCRIPTION
**Pull request by goozzik**
TP-2176
https://distribusion.tpondemand.com/entity/2176

### CODE REVIEW

File: lib/supa/commands/attributes.rb
Comment (Strech) [inline lib/supa/commands/attributes.rb:3]: In general, attributes should also have option like `hide_when_empty`. Maybe it's possible to implement it like multiple attribute commands 😉 ?

Comment (goozzik) [reply lib/supa/commands/attributes.rb:3]: @Strech It is possible but it was @dasnotme's request to implement it first without any options.

Comment (Strech) [reply lib/supa/commands/attributes.rb:3]: Yep, even that – it's still multiple singular 💃

Comment (goozzik) [reply lib/supa/commands/attributes.rb:3]: Added https://github.com/distribusion/supa/pull/21/commits/5eda8d3e83f7f9f3a193ef2c2a53489acc0de128. I had to change the interface of this command to make it work with options. Now it is

```ruby
attributes [:attribute_1, :attibute_2], { hide_when_nil: true }
```

**Comment by goozzik** [comment]
@Strech Thanks, please check again.

File: lib/supa/commands/attributes.rb
Comment (Strech) [reply lib/supa/commands/attributes.rb:3]: I don't see a reason why you change interface, please don't do that, you break Open-Close principle.
it doesn't make so much sense `name` it or `names`

File: spec/fixtures/article_representer.rb
Comment (Strech) [inline spec/fixtures/article_representer.rb]: Please do it like `attributes :title, :text`

File: lib/supa/builder.rb
Comment (Strech) [inline lib/supa/builder.rb]: Please don't do array because you can do it `*names`

**Review by Strech — CHANGES_REQUESTED** [review]
1. Don't break interface
2. No arrays in `attributes` usage, only implicit way

File: lib/supa/builder.rb
Comment (goozzik) [reply lib/supa/builder.rb]: @Strech It is not possible to do it like this:

builder_rb_ ___projects_dt_supa

We can do it with kvargs `def attributes(*names, options: {})`

Comment (Strech) [reply lib/supa/builder.rb]: yeah, because we have options at the end, but you can change `options` also ;)

Comment (ghost) [reply lib/supa/builder.rb]: 1) Tbh, I don't like exceptions of rules as people tend to add more and more if there's at least one already. We could refactor the code this way:
```ruby
COMMANDS.each do |command|
  klass = Supa::Commands.const_get(command.capitalize)

  define_method(command) do |*args, &block|
    options = args[-1].is_a?(Hash) ? args.pop : {}

    klass.new(@subject,
      representer: @representer,
      tree: @tree,
      args: args,
      options: options,
      &block).represent
  end
end
```

2) I'm not convinced we really need a new command, the same can be achieved by just calling `attribute` for every single argument of `attributes`, hence you don't have to implement the same functionaliy twice as well as break the interface of commands.
```ruby
def attributes(*args)
  options = args[-1].is_a?(Hash) ? args.pop : {}
  names = args

  names.each do |name|
    attribute(name, options)
  end
end
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: lib/supa/builder.rb
Comment (goozzik) [reply lib/supa/builder.rb]: I like what @dasnotme suggested so I've used it https://github.com/distribusion/supa/pull/21/commits/8cfbf8b92adf6d60944af5912e60c102b5409a4e. 

@Strech

**Comment by goozzik** [comment]
@Strech @dasnotme Thanks for the feedback, please continue.

File: lib/supa/builder.rb
Comment (Strech) [inline lib/supa/builder.rb]: Guys, looks like you are going to some nightmare direction

```ruby
def pew(*args, **kvargs)
  p args
  p kvargs
end
```

```bash
irb(main):005:0> pew(1,2,3)
[1, 2, 3]
{}
=> {}
irb(main):007:0> pew(1,2,3, a: 'LOL')
[1, 2, 3]
{:a=>"LOL"}
=> {:a=>"LOL"}
```

Comment (Strech) [reply lib/supa/builder.rb]: @goozzik Let's change the main interface from the old school, to ruby 2.x because it doesn't make sense to suffer without reason

1. Good new `**kvargs` -> `define_method(command) do |name, **options, &block|`
2. Let's have a normal `attributes` class with it's own behaviour
3. Let's implement `attributes` as multiple-times `attribute` without options hack, but just passing them into single case
4. No square brackets in interface, just don't count that option, no

Because now solution look dirty and hacky and don't follow the logic created before

Comment (goozzik) [reply lib/supa/builder.rb]: @Strech  https://github.com/distribusion/supa/pull/21/commits/b3bec0a71731cce71209df8634b70da153441b60 https://github.com/distribusion/supa/pull/21/commits/270b4a01a440646b18751d23e6084324fd646643

@dasnotme Please check the solution.

**Review by ghost — COMMENTED** [review]
I still believe in what I mentioned in https://github.com/distribusion/supa/pull/21#discussion_r103206966, but feel free to merge if you think otherwise.

**Review by ghost — APPROVED** [review]

**Comment by goozzik** [comment]
Then let's merge it because we will never settle this. I will merge after parent branch gets merged to master.

File: lib/supa/commands/attributes.rb
Comment (Strech) [inline lib/supa/commands/attributes.rb]: Please just pass all options you have to attribute, it's smart enough to choose

File: lib/supa/commands/attributes.rb
Comment (Strech) [inline lib/supa/commands/attributes.rb]: ✂️

**Review by Strech — CHANGES_REQUESTED** [review]
Despite small changes 👍

File: lib/supa/commands/attributes.rb
Comment (goozzik) [reply lib/supa/commands/attributes.rb]: @Strech It will allow to use all `attribute` command options. Some of the don't make sense to use when using `attributes` command.

Comment (Strech) [reply lib/supa/commands/attributes.rb]: Just pass them, `attribute` will decide how to work with them. It's not a responsibility of `attributes`. For instance then it will be possible to define single modifier to all and so on. And when we will need to add new one, we will not whitelist it

Comment (Strech) [reply lib/supa/commands/attributes.rb]: And attributes is a plural attribute, that's it

Comment (goozzik) [reply lib/supa/commands/attributes.rb]: done https://github.com/distribusion/supa/pull/21/commits/3202b313a1d5f8845af356a2e7f20a3981cffa37

File: lib/supa/commands/attributes.rb
Comment (goozzik) [reply lib/supa/commands/attributes.rb]: done https://github.com/distribusion/supa/pull/21/commits/3202b313a1d5f8845af356a2e7f20a3981cffa37

**Comment by goozzik** [comment]
@Strech Please check now.

**Review by Strech — APPROVED** [review]
Awesome 💠

**Pull request merged**
